### PR TITLE
DDCNL-3568 CySP COPE MVP

### PR DESCRIPTION
**Pull request by gianick**

### CODE REVIEW

File: test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    "return GateWayTimeout when GatewayTimeoutException is passed" in  {
+      val result = copeErrorHandling.errorWrapper(Future.failed(new GatewayTimeoutException("Gateway Timeout")))
+
+      status(result) shouldBe 504
+    }
+
+    "return BadGateway" when {
+      List(new BadGatewayException("BadGateway"),
+        new Upstream4xxResponse("4xx Response", 401, 502),
```
Comment (pascal22p) [inline test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala:61]: Why 401?

File: test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    "return Bad Request when BadRequestException is passed" in  {
+      val result = copeErrorHandling.errorWrapper(Future.failed(new BadRequestException("Bad Request")))
+
+      status(result) shouldBe 400
+      contentAsJson(result) shouldBe Json.toJson(ErrorGenericBadRequest("Upstream Bad Request. Is this customer below State Pension Age?"))
+    }
+
+    "return Forbidden" when {
+      "Upstream4xxResponse is returned with 422 status and NO_OPEN_COPE_WORK_ITEM message from DES" in  {
+        val result = copeErrorHandling.errorWrapper(Future.failed(new Upstream4xxResponse("NO_OPEN_COPE_WORK_ITEM", 422, 500)))
```
Comment (pascal22p) [inline test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala:82]: Why 500? Should it be 422?

File: test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+        status(result) shouldBe 403
+        contentAsJson(result) shouldBe Json.toJson(ExclusionCopeProcessing(inject[AppConfig]))
+      }
+
+      "Upstream4xxResponse is returned with 422 status and CLOSED_COPE_WORK_ITEM message from DES" in  {
+        val result = copeErrorHandling.errorWrapper(Future.failed(new Upstream4xxResponse("CLOSED_COPE_WORK_ITEM", 422, 500)))
```
Comment (pascal22p) [inline test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala:89]: Why 500? Should it be 422?

**Review by pascal22p — CHANGES_REQUESTED** [review]

File: test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala (hunk 2 above)
Comment (Tateni93) [reply test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala:82]: After looking at how the rawReads returns a 422 response in Bootstrap. We see that it returns an Upstream4xxResponse that reports as a 500. We have mimicked this in the test. Here's the code.
https://github.com/hmrc/http-verbs/blob/eb80b3de0187317920cf23a287cfb81fc55052d3/http-verbs-common/src/main/scala/uk/gov/hmrc/http/HttpErrorFunctions.scala#L48

File: test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala (hunk 3 above)
Comment (Tateni93) [reply test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala:89]: After looking at how the rawReads returns a 422 response in Bootstrap. We see that it returns an Upstream4xxResponse that reports as a 500. We have mimicked this in the test. Here's the code.
https://github.com/hmrc/http-verbs/blob/eb80b3de0187317920cf23a287cfb81fc55052d3/http-verbs-common/src/main/scala/uk/gov/hmrc/http/HttpErrorFunctions.scala#L48

File: test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala (hunk 1 above)
Comment (Tateni93) [reply test/uk/gov/hmrc/statepension/controllers/CopeErrorHandlingSpec.scala:61]: We wanted to mimic readsRaw in the test as errors come straight through we have tried to mimic readsRaw. A 400 would normally return BadRequestException,  404 is NotFoundException, 401 is the first one I thought that would fit.
https://github.com/hmrc/http-verbs/blob/eb80b3de0187317920cf23a287cfb81fc55052d3/http-verbs-common/src/main/scala/uk/gov/hmrc/http/HttpErrorFunctions.scala#L48

File: app/uk/gov/hmrc/statepension/controllers/ErrorHandling.scala [hunk 4]
```diff
@@ -16,31 +16,42 @@
...
+    func.recover {
+      case _: NotFoundException => NotFound(Json.toJson(ErrorNotFound))
       case e: GatewayTimeoutException => Logger.error(s"$app Gateway Timeout: ${e.getMessage}", e); GatewayTimeout
       case e: BadGatewayException => Logger.error(s"$app Bad Gateway: ${e.getMessage}", e); BadGateway
       case _: BadRequestException => BadRequest(Json.toJson(ErrorGenericBadRequest("Upstream Bad Request. Is this customer below State Pension Age?")))
+      case e: Upstream4xxResponse if appConfig.copeFeatureEnabled && e.upstreamResponseCode == 422 && e.message.contains("NO_OPEN_COPE_WORK_ITEM") =>
+        Forbidden(Json.toJson(ErrorResponses.ExclusionCopeProcessing(appConfig)))
+      case e: Upstream4xxResponse if appConfig.copeFeatureEnabled && e.upstreamResponseCode == 422 && e.message.contains("CLOSED_COPE_WORK_ITEM") =>
+        Forbidden(Json.toJson(ErrorResponses.ExclusionCopeProcessingFailed))
       case e: Upstream4xxResponse => Logger.error(s"$app Upstream4XX: ${e.getMessage}", e); BadGateway
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply app/uk/gov/hmrc/statepension/controllers/ErrorHandling.scala:53]: in this case DES is returning a 4xxException as the response we are expecting is coming as a 422. We need to capture these specifically and return a 403 as these are Exclusion cases. In short; the 422 is expected from DES and that is how they've decided to return an exclusion case to us. With regards to the regular handling of 4xxExclusions other than the 422 cases. We could probably tidy this up. Would your suggestion be simply to pass the Status coming from DES all the way or capture it a t the connector level and deal with it there?

[the remark this answers is not in the document]
Comment (pascal22p) [reply app/uk/gov/hmrc/statepension/controllers/ErrorHandling.scala:53]: Agreed off line on how to proceed

**Review by pascal22p — APPROVED** [review]

**Pull request merged**